File: studentassistant/src/main/java/ru/erdenian/studentassistant/activity/LessonsEditorActivity.java
package ru.erdenian.studentassistant.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.PagerTabStrip;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ScrollView;
import android.widget.Spinner;

import ru.erdenian.studentassistant.R;
import ru.erdenian.studentassistant.adapter.ScheduleEditorPagerAdapter;
import ru.erdenian.studentassistant.schedule.OnScheduleUpdateListener;
import ru.erdenian.studentassistant.schedule.ScheduleManager;

public class LessonsEditorActivity extends AppCompatActivity implements
        View.OnClickListener,
        AdapterView.OnItemSelectedListener,
        OnScheduleUpdateListener {

    static final String SEMESTER_ID = "semester_id";

    private long semesterId = -1;
    private int semesterIndex = -1;

    private int savedPage = 0;

    private ViewPager viewPager;
    private ScrollView scrollView;
    private ScheduleEditorPagerAdapter pagerAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_lessons_editor);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar_with_spinner);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowTitleEnabled(false);

        Spinner spEditTypes = (Spinner) findViewById(R.id.toolbar_with_spinner_spinner);
        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, R.layout.spinner_item_semesters,
                getResources().getStringArray(R.array.activity_lessons_editor_edit_types));
        adapter.setDropDownViewResource(R.layout.spinner_dropdown_item_semesters);
        spEditTypes.setAdapter(adapter);
        spEditTypes.setOnItemSelectedListener(this);

        viewPager = (ViewPager) findViewById(R.id.view_pager);
        PagerTabStrip pagerTabStrip = (PagerTabStrip) findViewById(R.id.view_pager_pager_tab_strip);
        pagerTabStrip.setTextColor(ContextCompat.getColor(this, R.color.colorPrimary));
        pagerTabStrip.setTabIndicatorColorResource(R.color.colorPrimary);

        scrollView = (ScrollView) findViewById(R.id.scroll_view);

        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.activity_lessons_editor_add_lesson);
        fab.setOnClickListener(this);

        semesterId = getIntent().getLongExtra(SEMESTER_ID, -1);
    }

    @Override
    protected void onStart() {
        super.onStart();
        ScheduleManager.INSTANCE.setOnScheduleUpdateListener(this);
        onScheduleUpdate();
    }

    @Override
    public void onScheduleUpdate() {
        if ((semesterIndex == -1) || (semesterIndex >= ScheduleManager.INSTANCE.getSemesters().size()) ||
                (ScheduleManager.INSTANCE.getSemesters().asList().get(semesterIndex).getId() != semesterId)) {
            semesterIndex = ScheduleManager.INSTANCE.getSemesterIndex(semesterId);

            if (semesterIndex == -1) {
                finish();
                return;
            }
        }

        if (pagerAdapter != null) {
            savedPage = viewPager.getCurrentItem();
        }

        pagerAdapter = new ScheduleEditorPagerAdapter(getSupportFragmentManager(), ScheduleManager.INSTANCE.getSemesters().asList().get(semesterIndex));
        viewPager.setAdapter(pagerAdapter);
        viewPager.setCurrentItem(savedPage, false);
        savedPage = 0;

        // TODO: 13.11.2016 добавить заполнение списка пар по датам
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_lessons_editor, menu);
        //UiUtils.INSTANCE.colorMenu(this, menu);
        return true;
    }

    @Override
    public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
        viewPager.setVisibility((i == 0) ? View.VISIBLE : View.GONE);
        scrollView.setVisibility((i == 1) ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
            case R.id.menu_lessons_editor_edit_semester:
                Intent intent = new Intent(this, SemesterEditorActivity.class);
                intent.putExtra(SemesterEditorActivity.SEMESTER_ID, semesterId);
                startActivity(intent);
                break;
            case R.id.menu_lessons_editor_delete_semester:
                ScheduleManager.INSTANCE.removeSemester(semesterIndex);
                break;
            default:
                Log.wtf(this.getClass().getName(), "Неизвестный id: " + item.getItemId());
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.activity_lessons_editor_add_lesson:
                break;
            default:
                Log.wtf(this.getClass().getName(), "Неизвестный id: " + v.getId());
                break;
        }
    }
}

File: studentassistant/src/main/kotlin/ru/erdenian/studentassistant/activity/LessonsEditorActivity.kt
package ru.erdenian.studentassistant.activity

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import kotlinx.android.synthetic.main.activity_lessons_editor.*
import kotlinx.android.synthetic.main.scroll_view.*
import kotlinx.android.synthetic.main.toolbar_with_spinner.*
import kotlinx.android.synthetic.main.view_pager.*
import org.jetbrains.anko.alert
import org.jetbrains.anko.startActivity
import ru.erdenian.studentassistant.R
import ru.erdenian.studentassistant.adapter.SchedulePagerAdapter
import ru.erdenian.studentassistant.extensions.getCompatColor
import ru.erdenian.studentassistant.extensions.setColor
import ru.erdenian.studentassistant.schedule.OnScheduleUpdateListener
import ru.erdenian.studentassistant.schedule.ScheduleManager

class LessonsEditorActivity : AppCompatActivity(),
    View.OnClickListener,
    AdapterView.OnItemSelectedListener,
    OnScheduleUpdateListener {

  private val semesterId: Long by lazy {
    intent.getLongExtra(SEMESTER_ID, -1L).takeIf { it != -1L } ?: throw IllegalStateException("Не передан id семестра")
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_lessons_editor)

    setSupportActionBar(toolbar_with_spinner)
    supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    //supportActionBar!!.setDisplayShowTitleEnabled(false)

    val adapter = ArrayAdapter(this, R.layout.spinner_item_semesters,
        resources.getStringArray(R.array.lesson_repeat_types))
    adapter.setDropDownViewResource(R.layout.spinner_dropdown_item_semesters)
    toolbar_with_spinner_spinner.adapter = adapter
    toolbar_with_spinner_spinner.onItemSelectedListener = this

    toolbar_with_spinner_spinner.visibility = View.GONE

    view_pager_pager_tab_strip.setTextColor(getCompatColor(R.color.colorPrimary))
    view_pager_pager_tab_strip.setTabIndicatorColorResource(R.color.colorPrimary)

    activity_lessons_editor_add_lesson.setOnClickListener(this)
  }

  override fun onStart() {
    super.onStart()
    ScheduleManager.addOnScheduleUpdateListener(this)
    onScheduleUpdate()
  }

  override fun onScheduleUpdate() {
    val semester = ScheduleManager.getSemester(semesterId)

    if (semester != null) {
      val page = view_pager.currentItem
      view_pager.adapter = SchedulePagerAdapter(supportFragmentManager, semester, true)
      view_pager.currentItem = page

      // TODO: 13.11.2016 добавить заполнение списка пар по датам
    } else {
      finish()
    }
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.menu_lessons_editor, menu)
    menu.setColor(getCompatColor(R.color.action_bar_icons_color))
    return true
  }

  override fun onItemSelected(adapterView: AdapterView<*>, view: View?, i: Int, l: Long) {
    view_pager.visibility = if (i == 0) View.VISIBLE else View.GONE
    scroll_view.visibility = if (i == 1) View.VISIBLE else View.GONE
  }

  override fun onNothingSelected(parent: AdapterView<*>) = Unit

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    when (item.itemId) {
      android.R.id.home -> finish()
      R.id.menu_lessons_editor_edit_semester -> startActivity<SemesterEditorActivity>(SEMESTER_ID to semesterId)
      R.id.menu_lessons_editor_delete_semester -> {
        fun remove() {
          ScheduleManager.removeSemester(semesterId)
          finish()
        }

        alert(R.string.activity_lessons_editor_alert_delete_message) {
          positiveButton(R.string.activity_lessons_editor_alert_delete_yes) { remove() }
          negativeButton(R.string.activity_lessons_editor_alert_delete_no)
        }.show()
      }
      else -> throw IllegalArgumentException("Неизвестный id: ${item.itemId}")
    }
    return super.onOptionsItemSelected(item)
  }

  override fun onClick(v: View) {
    when (v.id) {
      R.id.activity_lessons_editor_add_lesson ->
        startActivity<LessonEditorActivity>(
            SEMESTER_ID to semesterId,
            WEEKDAY to view_pager.currentItem + 1
        )
      else -> throw IllegalArgumentException("Неизвестный id: ${v.id}")
    }
  }
}
